users-flask/app.py

```diff
@@ -0,0 +1,162 @@
+    app.logger.exception(e)
+    data = CustomException(str(e)).json()
+    res = ResponseBodyJSON(False, data).json()
+    return jsonify(res), 500
+
+
+@app.route('/users')
+def find_all_users():
+    args = request.args
+
+    milestone_id = args.get('milestone_id')
+    if milestone_id is not None:
+        offsets = db.session.execute(db.select(Offset).where(Offset.milestone_id == milestone_id)).scalars().all()
+        data = [offset.buyer.json() for offset in offsets]
+        res =  ResponseBodyJSON(True, data).json()
+        return jsonify(res), 200
+
+    else:
+        users = db.session.execute(db.select(User)).scalars().all()
+        data = [user.json() for user in users]
+        res =  ResponseBodyJSON(True, data).json()
+        return jsonify(res), 200
+    
+@app.route("/users/<uuid:id>")
+def find_user_by_id(id):
+    user = db.session.execute(db.select(User).where(User.id == id)).scalars().first()
+
+    if user is None:
+        abort(404, description=f"User {str(id)} not found.")
+    
+    data = user.json()
+    res = ResponseBodyJSON(True, data).json()
+    return jsonify(res), 200
+
+@app.route("/users", methods=['POST'])
+def create_user():
+    body = request.get_json()
+
+    user = User(**body)
+    db.session.add(user)
+    db.session.commit()
+    
+    data = user.json()
+    res = ResponseBodyJSON(True, data).json()
+    return jsonify(res), 201
+    
+
+@app.route("/users/verifiers")
+def find_verifiers():
+    verifiers = db.session.execute(db.select(User).where(User.is_verifier)).scalars().all()
+    data = [v.json() for v in verifiers]
+    res = ResponseBodyJSON(True, data).json()
+    return jsonify(res), 200
+
+
+@app.route("/users/<uuid:id>/offset", methods=['POST'])
+def create_user_offset(id):
+    body = request.get_json()
+
+    user = db.session.execute(db.select(User).where(User.id == id)).scalars().first()
+    if user is None:
+        abort(404, description=f"User {str(id)} not found.")
+
+    offset = Offset(**body, buyer=user)    
+    db.session.add(offset)
+    db.session.commit()
+    
+    data = offset.json()
+    res = ResponseBodyJSON(True, data).json()
+    return jsonify(res), 201
+
+
+@app.route("/users/<uuid:id>/offset/<string:payment_id>", methods=['PATCH'])
+def update_buyer_offset(id, payment_id):
+    """
+    If status is confirmed, 
+    1. update the offset status to 'confirmed'
+    2. the buyer's footprint is increased.
+    else if status is 'refund', only the offset status is updated as 'refund'.
+    """
+    
+    buyer: User or None = db.session.execute(db.select(User).where(User.id == id)).scalars().first()
+    if buyer is None:
+        abort(404, description=f"User {str(id)} not found.")
+        
+    body = request.get_json()
+    
+    status = body['status']
+    if status not in OffsetStatus.values():
+        abort(400, description=f"Invalid status '{status}'. Status can only be the following: {OffsetStatus.values()}")
```

Comment:
Test data for PATCH: update buyer offset change to:
`
{
    "status": 'pending' | 'confirmed' | 'refund'

`